attestation-agent/attester/src/tsm_report/mod.rs

```diff
@@ -0,0 +1,125 @@
+// Copyright (c) 2024 Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+use log::error;
+use std::path::{Path, PathBuf};
+use tempfile::tempdir_in;
+use thiserror::Error;
+
+const TSM_REPORT_PATH: &str = "/sys/kernel/config/tsm/report";
+pub const TSM_REPORT_PROVIDER_TDX: &str = "tdx_guest";
+pub const TSM_REPORT_PROVIDER_SEV: &str = "sev_guest";
+
+#[derive(Error, Debug)]
+pub enum TsmReportError {
+    #[error("Failed to create TSM Report path: {0}")]
+    Open(#[from] std::io::Error),
+    #[error("Failed to access TSM Report attribute: {0}: {1}")]
+    Access(&'static str, #[source] std::io::Error),
+    #[error("TSM Report inblob write conflict: generation {0}, expected 1")]
+    InblobConflict(u32),
+    #[error("Failed to generate TSM Report: missing inblob (len=0)")]
+    InblobLen,
+}
+
+pub enum TsmReportProvider {
+    Tdx(Vec<u8>),
+    Sev(u8, Vec<u8>),
+}
+
+pub struct TsmReportPath {
+    path: PathBuf,
+    pub provider: String,
+}
+
+impl TsmReportPath {
+    pub fn open() -> Result<Self, TsmReportError> {
+        let p = tempdir_in(TSM_REPORT_PATH).map_err(TsmReportError::Open)?;
+
+        let path = p.into_path();
+
+        let provider = match std::fs::read_to_string(path.as_path().join("provider")) {
+            Ok(tsm) => tsm.trim_matches('\n').to_string(),
+            Err(e) => {
+                let _ = std::fs::remove_dir(path.as_path())
+                    .map_err(|r| log::error!("Failed to remove TSM Report directory: {}", r));
+                return Err(TsmReportError::Access("provider", e));
+            }
+        };
+
+        Ok(Self { path, provider })
+    }
+    pub fn close(&self) {
+        let _ = std::fs::remove_dir(self.path.as_path())
```

Comment:
Wasn't this a temporary directory?

Reply:
Yes but the automatic removal is broken (details: with configfs, only mkdir/rmdir works but `tempdir` uses `remove_dir_all()` which first tries to remove files in the tmpdir and that returns an error. so, `tempdir` just silently ignores the error and leaks the tempdir)